---
title: 'Data analysis: Project 2 Group 4'
author: "Anna Muench, Ann-Sophie Kroell, Jana Braunger"
subtitle: ' Molecular Biotechnology, 4th term, Summer 2019'
output:
  html_document:
    hightlight: monochrome
    number_sections: no
    theme: yeti
    toc: yes
    toc_depth: 3
  pdf_document:
    toc: yes
    toc_depth: '3'
---
# **Broad Analysis**

```{r, include = FALSE}
# load required packages
library(knitr)
```

```{r setup, include = FALSE} 
# code and output not included in final file
knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(fig.width = 9, fig.height = 6, fig.align = "center")
```

<div style = "text-align:justify">
The project is based on the data generated in the NCI Transcriptional Pharmacodynamics Workbench (TPW) data and is divided into a broad and specific analysis. First of all, a broad analysis (milestone 1) containing all samples of the NCI60 panel with 15 different cancer drugs is performed. In the specific analysis we focus on the drug erlotinib, which is an inhibitor of EGFR. Four milestones are defined to analyse which pathways are mostly regulated due to the drug treatment. The second milestone includes finding the cell lines which have the strongest fold change and the third one to identify biomarkers, defined as the genes, which are most regulated, significant and best correlating with drug sensitivity. In milestone 4 the correlation between the drug sensitivity of the various cell lines (GI50) and the EGR1 expression is analysed. In the last part, the results of the previous milestones are put together to show the key results of our project.

### Load data
Firstly, the data generated from the NCI TPW is downloaded and the various tables are saved as data frames. 

```{r }
NCI_TPW_gep_treated <- readRDS(url("https://ndownloader.figshare.com/files/14720180?private_link=db1411debcfbe2618d2f"))
NCI_TPW_gep_untreated <- readRDS(url("https://ndownloader.figshare.com/files/14720183?private_link=db1411debcfbe2618d2f"))
NCI_TPW_metadata <- read.delim("https://ndownloader.figshare.com/files/14720186?private_link=db1411debcfbe2618d2f")
NegLogGI50 <- readRDS(url("https://ndownloader.figshare.com/files/ 14720210?private_link=074e0120fe5e682f3d14"))
CCLE_basalexpression <- readRDS(url("https://ndownloader.figshare.com/files/14770127?private_link=fc0c71246dc192982a3c"))
CCLE_copynumber <- readRDS(url("https://ndownloader.figshare.com/files/14770130?private_link=fc0c71246dc192982a3c"))
CCLE_mutations <- readRDS(url("https://ndownloader.figshare.com/files/14770133?private_link=fc0c71246dc192982a3c"))
cellline_annotation <-read.delim("https://ndownloader.figshare.com/files/14768981?private_link=efb6a529eaf2d4dc6432")
drug_annotation <- read.delim("https://ndownloader.figshare.com/files/14768984?private_link=efb6a529eaf2d4dc6432")
```

##   _**1. Milestone: Broad analysis**_

The first part of our project is a broad analysis to get an overview of the data generated in the NCI Transcriptional Pharnacodynamics Workbench (TPW). The distribution of the data is analyzed with a density plot, as well as with a boxplot to show, whether the data is already normalized or normalization is needed for further analysis. The data of the TPW contains gene expression values for 819 samples and 13299 genes. For each of the cancer cell lines the normal gene expression, as well as the gene expression after drug treatment with one of 15 drugs was measured. In a principal component analysis it is tested, whether there are clusters of samples corresponding to the same cancertype or samples treated with the same drug by coloring according to these two attributes. In the last step of the overall analysis we search for genes, which were most regulated through the therapeutic cancer treatment.  

### Data preparation and annotation

After loading the data of the TPW, we need some preparation and annotation of the data to perform our analysis. To show the effect of a drug on the gene expression, the log2 fold change, which is a measurement for the quantitative change due to drug treatment, is used. Since the gene expression values are already logarithmic, the fold change can be calculated by substracting the untreated values (as a control of normal gene expression for each gene in each sample) from the treated ones. 
  
#### Calculate fold change due to drug treatment  

```{r }
fold_changes <- NCI_TPW_gep_treated - NCI_TPW_gep_untreated
fold_changes <- as.data.frame(fold_changes)
```

    
#### Renaming of cell line SK-MEL-2 
One problem, which occured at the first structuring of the data by using the grep() function was that the cell line name SK-MEL-2 is part of cell line SK-MEL_28. To solve this we rename that cell line to SK-MEL-2_.  

```{r }
# SK-MEL-2_ is added as new factor 
levels(cellline_annotation$Cell_Line_Name) <- c(levels(cellline_annotation$Cell_Line_Name), 
                                                "SK-MEL-2_")
cellline_annotation[33, 1] <- "SK-MEL-2_"
# delete level SK-MEL-2 (otherwise we would have 62, instead of 61 levels)
cellline_annotation$Cell_Line_Name <- factor(as.character(
  cellline_annotation$Cell_Line_Name))
```
  
#### Renaming of cancertype Non-Small Cell Lung 
To make the legend of the following plots smaller and clearer, we abbreviate Non-Small Cell Lung with NSCL.
```{r}
# NSCL is added as new factor 
levels(cellline_annotation$Cancer_type) <- c(levels(cellline_annotation$Cancer_type), 
                                                "NSCL")
cellline_annotation[grep ("Non-Small Cell Lung", cellline_annotation$Cancer_type), "Cancer_type"] <- "NSCL"
# delete level Non-Small Cell Lung (otherwise we would have 16, instead of 15 levels)
cellline_annotation$Cancer_type <- factor(as.character(
  cellline_annotation$Cancer_type))
```
  
#### Annotation of all sample names  
A dataframe ("annotation") is created, which contains for each sample name (in 819 rows) the drug, cell line and cancertype. This annotation is later used for labeling and coloring of our plots.     

**1. Drug**   
For finding the drug a certain sample is treated with, the function grepl is used which shows whether a drug listed in the drug_annotation is contained in the sample name (colnames of fold_changes).
```{r }
sample_drug <- as.data.frame(sapply(levels(drug_annotation$Drug), grepl, 
                                    colnames(fold_changes), ignore.case = TRUE))
  # creates table with TRUE and FALSE for each sample (819 rows) and drug (15 columns)
rownames(sample_drug) <- colnames(fold_changes) # sample names as rownames
drugs <- as.vector(apply(sample_drug, 1, function(x){ # for each row of the sample_drug dataframe (=samples)
  # save the drug, which is the columnname in the column of the TRUE
  colnames(sample_drug[which(x)]) 
}))
```

**2. Cell line**  
The cell line is extracted from the sample name in a similar way as the drug. The two vectors with the drug and cell line of each sample are then combined in one data frame.
```{r }
sample_cellline <- as.data.frame(sapply(levels(cellline_annotation$Cell_Line_Name), grepl, 
                                        colnames(fold_changes), ignore.case = TRUE)) 
  # creates table with TRUE and FALSE for each sample (819 rows) and cell line (61 columns)
rownames(sample_cellline) <- colnames(fold_changes) #sample names as rownames
cellline <- as.vector(unlist(apply(sample_cellline, 1, function(x){ 
  # for each row of the sample_drug dataframe (=samples)
  #save the cellline, which is the columnname in the column of the TRUE
  colnames(sample_cellline[which(x)])
})))

# annotation with drugs and cell lines in columns for each of the samples (819 rows)
annotation <- cbind("Drug" = drugs, "Cellline" = cellline)
rownames(annotation) <- colnames(fold_changes)
```

**3. Cancertype**  
In the cellline_annotation each cell line is assigned to one of seven different cancertypes, which is added as a third column in our annotation. 
```{r }
cancertype <- sapply(annotation[, 2], function(x){ 
  # 2nd column of annotation contains cell line name of samples
  # save the cancertype the cell line belongs to 
  # by taking the word in cellline_annotation in column Cancer_type and in the row, where the cell line x stands
  cellline_annotation$Cancer_type[cellline_annotation$Cell_Line_Name == x]
})
cancertype <- as.vector(unlist(cancertype)) #to get a vector instead of a list of the cancertypes per sample

# add cancertype in annotation
annotation <- cbind(annotation, "Cancertype" = cancertype)
# remove all variables which are no longer needed
rm(drugs, sample_drug, cellline, sample_cellline, cancertype)
```

#### Preparation for Coloring    
For coloring according to drug or cancertype a color vector for each of them is created which assigns each drug or each cancertype a color. These color vectors are used for coloring of the plots and creating the corresponding legends. For this purpose we searched for a color palette containing 15 colors, which are easy to distinguish. However, we only found some like RColorBrewer, which had no palette containing at least 15 distinguishable colors. Therefore we just define our own color_palette by using the names of easily distinguishable colors. 

**1. Coloring according to drug (color_vector_all_drugs)**
```{r }
# define a color palette with 15 chosen colors
color_palette_drug <- c("aquamarine", "brown", "forestgreen", "slategrey", 
                        "chartreuse", "darkgoldenrod1", "cadetblue","purple", 
                        "firebrick1", "deepskyblue", "gold", "violetred4", 
                        "deeppink", "plum2", "blue" )
# assign each color one of the 15 drugs
names(color_palette_drug) <- levels(drug_annotation$Drug)
  
# create vector containing a color name for each sample according to drug (length 819)
color_vector_drug <- sapply(rownames(annotation), function(x){ # for all samples x
  unname(color_palette_drug[annotation[x, 1]]) # first column of annotation contains drugs
})
```

**2. Coloring according to cancertype (color_vector_cancertype)**
```{r }
# define a color palette with 9 chosen colors
color_palette_cancertype <- c("aquamarine", "brown", "forestgreen", "chartreuse", 
                              "darkgoldenrod1", "cadetblue","purple", 
                              "firebrick1", "deepskyblue")

# each color is assigned the name of the 9 cancertypes
names(color_palette_cancertype) <- levels(cellline_annotation$Cancer_type)

# create vector containing a color name for each sample according to cancertype (length 819)
color_vector_cancertype <- sapply(rownames(annotation), function(x){
  unname(color_palette_cancertype[annotation[x, 3]]) #3rd columns of annotation contains cancertype 
})
```

### Distribution of gene expression values for "untreated" and "treated"
To show the distribution of all gene expression values, a **density plot** is drawn. The black line contains all values measured for control samples (untreated). In red the distribution of the gene expression of all samples treated with 15 drugs is shown.
```{r }
plot(density(NCI_TPW_gep_untreated), 
     "Density plot of gene expression", 
     cex.main = 1.2) # sets size of main
lines(density(NCI_TPW_gep_treated), col = "red")
legend("topright", 
       legend = c("untreated", "treated"), 
       col = c("black", "red"), # colors
       pch = 15) # filled square to show color in legend
```

As expected, there can be hardly seen any difference between both curves, so the distribution is nearly the same. The main reason for that is that the gene expression of most of the 13299 genes does hardly or not at all change due to the drug. Ideally the drug should only affect genes, which have an impact specifically on cancer cells and allow them to divide endlessly. On the other side, other genes should not be affected to remain a functional cell. Furthermore, an upregulation of some genes and a downregulation of others could compensate each other in the density plot. 

### Gene expression profile for untreated cell lines

In a next step the gene expression profile of each untreated sample is visualized in a **boxplot** to look, whether the expression profiles look the same over all samples or whether normalization is needed. 

```{r warning = FALSE }
# par used to set graphical parameters
# mar gives the margin size
# xpd = "TRUE" to plot legend outside of the plot
par(mar = c(4, 4, 4, 7.5), xpd = "TRUE")
boxplot(NCI_TPW_gep_untreated, 
        xaxt = "n", # removes labels on x-axis
        ylab = "Gene expression profile", 
        vertical =  T, 
        cex.main = 1.2,
        main = "Gene expression profile of untreated NCI60 cell lines")
title(xlab = "Samples", line = 1.0) # used to create a xlab nearer to the axis
```

In the boxplot differences occuring after a certain number of samples can be observed. An explanation for that could be that the gene expression of the samples was measured at different points of time or even at different laboratories. This raises the question, whether these batches match with the 15 drugs these control measurements of untreated expressions were made. Therefore, the boxes are **colored according to the drug** the control was used for.   

```{r warning = FALSE}
par(mar = c(4, 4, 4, 7.5), xpd = "TRUE")
boxplot(NCI_TPW_gep_untreated, 
        xaxt = "n", # removes labels on x-axis
        ylab = "Gene expression profile", 
        vertical =  T, 
        main = "Gene expression profile of untreated NCI60 cell lines", 
        cex.main = 1.2,
        boxcol = color_vector_drug)
title(xlab = "Samples", line = 1.0) # used to create a xlab nearer to the axis
legend("topright", 
       legend = names(color_palette_drug), 
       col = color_palette_drug,
       inset = c(-0.22, 0), # moves legend to the right
       pch = 19) # type of points in legend
```

Since the batches exactly match the different drugs for which the untreated expression is used, **normalization** is necessary if the values over the various drugs should be comparable. 

```{r warning = FALSE}
# each sample should have mean 0 and sd 1 in the untreated and treated table
# this is achieve by substracting the mean and dividing through the standard deviation of the sample
untreated_normalized <- apply(NCI_TPW_gep_untreated, 2, function(x){ 
  # for all samples (columns in NCI_TPW_gep_untreated)
  (x - mean(x)) / sd(x)
})
treated_normalized <- apply(NCI_TPW_gep_treated, 2, function(x){
  (x - mean(x)) / sd(x)
})

# the fold change is calculated from the normalized data
FC_normalized <- treated_normalized - untreated_normalized

# boxplot of normalized untreated values
par(mar = c(4, 4, 4, 7.5), xpd = "TRUE")
boxplot(untreated_normalized, 
        xaxt = "n", # removes labels on x-axis
        ylab = "Gene expression profile", 
        vertical =  T, 
        main = "Normalized gene expression profile of untreated NCI60 cell lines",
        cex.main = 1.2, 
        boxcol = color_vector_drug)
title(xlab = "Samples", line = 1.0) # used to create a xlab nearer to the axis
legend("topright",
       legend = names(color_palette_drug), 
       col = color_palette_drug, 
       pch = 19, # type of points in legend
       inset = c(-0.22, 0)) # moves legend to the right
```

### Principal Component Analysis (PCA) 
A principal component analysis is used for dimensionality reduction. With these technique it is possible to depict most of the variance observed in the gene expression changes due to drug treatment (foldchange) over all samples. The points are colored firstly according to drug and secondly according to cancertype to see whether there are corresponding clusters. 

#### Coloring according to drug
```{r warning = FALSE, fig.height = 10, message = FALSE}
library(factoextra)

# prcomp calculates principal components
pca <- prcomp(FC_normalized)

# set graphical parameters
# mfrow: combine multiple plots, here 2 rows and 1 column
par(mar = c(4, 4, 4, 7.5), mfrow = c(2, 1))

# PC1 and PC2
plot(pca$rotation[,1], 
     pca$rotation[,2], 
     col = color_vector_drug, 
     pch = 19, 
     # 2nd column of get_eigenvalue contains percentage of explained variance
     # round(x, digits to round): here round explained variance to one digit
     xlab = paste("PC1 (", round(get_eigenvalue(pca)[1,2], 1), "%)"), 
     ylab = paste("PC2 (", round(get_eigenvalue(pca)[2,2], 1), "%)"))

# create legend centered over both plots on the right side
legend("topright", 
       legend = names(color_palette_drug), 
       col = color_palette_drug, 
       pch = 19, 
       xpd = "TRUE",
       inset = c(-0.22, 0.7)) #move legend to the right and down

# Title for both plots together: mtext = margin text, side = 3 (upside)
mtext("PCA of FC colored according to drug", 
      side = 3, 
      line = -2.5, # giving the position as a margin line
      cex = 1.2,
      font = 2, # bold
      outer = TRUE)

# PC2 and PC3
plot(pca$rotation[,2], 
     pca$rotation[,3], 
     col = color_vector_drug, 
     pch = 19, 
     xlab = paste("PC2 (", round(get_eigenvalue(pca)[2,2], 1), "%)"), 
     ylab = paste("PC3 (", round(get_eigenvalue(pca)[3,2], 1), "%)"))

```

In the PCA plot it can be seen that the cell lines, treated with the same drug, cluster in certain areas. Especially the fold change of the cell lines treated with bortezomib, doxorubicin and vorinostat cluster further apart from the rest. The samples of erlotinib, on which we will focus in the second part are mostly clustered in the middle of the plot. 

#### Coloring according to cancertype
```{r warning = FALSE, fig.height = 10}
# set graphical parameters
# mfrow: combine multiple plots, here 2 rows and 1 column
par(mar = c(4, 4, 4, 7), mfrow = c(2, 1))

# PC1 and PC2
plot(pca$rotation[,1], 
     pca$rotation[,2], 
     col = color_vector_cancertype, 
     pch = 19, 
     # get_eigenvalue is used to add percentage of explained variance in xlab and ylab
     # 2nd column of get_eigenvalue contains percentage of explained variance
     # round(x, digits to round): here round explained variance to one digit
     xlab = paste("PC1 (", round(get_eigenvalue(pca)[1, 2], 1), "%)"), 
     ylab = paste("PC2 (", round(get_eigenvalue(pca)[2, 2], 1), "%)"))

# create legend centered over both plots on the right side
legend("topright",
       legend = names(color_palette_cancertype), 
       col = color_palette_cancertype, 
       pch = 19, 
       xpd = "TRUE",
       inset = c(-0.19, 0.9))

# PC2 and PC3
plot(pca$rotation[,2], 
     pca$rotation[,3], 
     col = color_vector_cancertype, 
     pch = 19, 
     xlab = paste("PC2 (", round(get_eigenvalue(pca)[2,2], 1), "%)"), 
     ylab = paste("PC3 (", round(get_eigenvalue(pca)[3,2], 1), "%)"))

# Title for both plots together: mtext = margin text, side = 3 (upside)
mtext("PCA of FC colored according to cancertype", 
      side = 3, 
      line = -2.5, # giving the position as a margin line
      cex = 1.2,
      font = 2, # bold
      outer = TRUE)

```

The colors showing which cancertype a cell line belongs to seem rather random distributed in the PCA plot. No clustering between the cell lines beeing part of the same cancertype can be observed. This means that the drug changes the gene expression in a similar way in all samples, whereas the cancertype has not that much impact on how the gene expression is changed.

### Most regulated genes 
In the last part of the broad analysis, we analyse which genes are most regulated over all drug treatments. For that the mean fold change for each gene over all samples is calculated and the 10 genes with the highest mean are shown in a **barplot**. The mean fold change is calculated from the absolute numbers of the fold change, since otherwise positive and negative values would compensate each other and the mean would probably be around zero.  
```{r}
# calculating the mean FC over positive FC values
mean_FC_abs <- apply(abs(fold_changes), 1, mean)
mean_FC_abs <- sort(mean_FC_abs, decreasing = TRUE)
mean_FC_abs <- as.data.frame(cbind(Genes = names(mean_FC_abs[1:10]), abs_foldchange = mean_FC_abs[1:10]))
mean_FC_abs$abs_foldchange <- as.numeric(as.character(mean_FC_abs$abs_foldchange))

# create barplot
ggplot(data = mean_FC_abs, aes(x = Genes, y = abs_foldchange)) +
    geom_bar(stat = "identity")+
    ylab("mean fold change over all samples per gene")+
    ggtitle("Genes with highest mean in absolute FC")+
    theme(plot.title = element_text(size = 13, face = "bold", hjust = 0.5))+ # bold and centered title
    scale_x_discrete(limits = rev(rownames(mean_FC_abs)))+ # sort cell lines by increasing values
    coord_flip()

```
The barplot shows that ATF3 has the highest mean fold change, which means that this gene is regulated the most over all cancer treatments. ATF3 (activating transcription factor 3) is an adaptive response gene, which means that it regulates the adaption to extra- or intracellular changes. Furthermore, ATF3 regulates the cell cycle and apoptosis, so it plays a crucial role in cancer progression (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2783469/).
  Other genes like CXCL8, which is part of the CXC chemokine family and a mediator of the inflammatory response (https://www.genecards.org/cgi-bin/carddisp.pl?gene=CXCL8), GDF15 (growth differentiation factor 15, https://www.genecards.org/cgi-bin/carddisp.pl?gene=GDF15) or DDIT3 (DNA damage inducible transcript 3, https://www.genecards.org/cgi-bin/carddisp.pl?gene=DDIT3), are also associated with cancer and should change the expression due to drug treatment. EGR1 (early growth response 1) is a transcription factor and is upregulated by mutant EGFR (https://www.ncbi.nlm.nih.gov/pubmed/19414352), which is the target of the drug erlotinib we focused on our specific analysis.

To visualize the distribution of the fold change of these highly regulated genes a **boxplot** over all samples is created.
```{r}
# FC_samples_with_highest_mean_FC contains the gene expression of the 20 biomarkers (20 columns) of all samples(819 rows) 
FC_samples_with_highest_mean_FC <- data.matrix(as.data.frame(sapply(rownames(mean_FC_abs), function(x){
  fold_changes[which(x == rownames(fold_changes)),]
})))

# boxplot showing the distribution of all values measured for the most regulated genes
par(mar = c(8, 2, 2, 2))
boxplot(FC_samples_with_highest_mean_FC, 
      ylab = "foldchange", 
      main = "Boxplot of foldchange of the genes with highest mean FC", 
      cex.main = 1.2,
      las = 2,
      outline = FALSE) 
```

It can be seen that the distribution of all of the highly regulated genes varies a lot over the different samples. All of these highly regulated genes are upregulated in some samples, while they are downregulated in others. Since, most of these genes play a role in pathways related with cancer, we would have expected that they are regulated in the same direction in the various cancercells by the various drugs, which is not the case. Nevertheless, more than 75% of the samples per gene have a positive fold change and only a few samples are regulated in the other direction.


# **Specific analysis: Erlotinib **

##   _**2. Milestone: Identify most affected cell lines**_

<div style = "text-align:justify">
The aim of the second milestone is to identify cell lines, which are highly affected by erlotinib treatment using three different methods. Firstly, we select 15 cell lines, which have the highest variance and have therefore the highest changes in regulation between the different genes. Additionally, we perform a PCA to select those cell lines, which differ most from the other cell lines and verify the previously found cell lines. Afterwards, we draw two different heatmaps. The first one contains the genes used in the supplementary material of the NCI TPW paper. For the second heatmap, we use the database PROGENy to get more precise results. With the heatmaps, we select those cell lines, which form a cluster with EGFR pathway downregulation. To sum up, we compare the three different methods using a Venn plot. We thereby select those cell lines, which appear in every of our three methods.

### Data preparation 

Before we start with our erlotinib specific analysis, we have to prepare our data. Firstly, we generate a new matrix that only contains erlotinib treated samples. This is done separately for our treated "e_treated" and untreated "e_untreated" data. Afterwards, we normalise our data sets and rename the columns with the cell line name instead of the full sample name. In our last preparation step, we calculate the foldchange and name the new matrix "e_foldchange."

```{r }
# new matrix only with samples treated with erlotinib  (e = erlotinib)
e_treated <- NCI_TPW_gep_treated[ ,grep ("erlotinib", colnames(NCI_TPW_gep_treated))]
e_untreated <- NCI_TPW_gep_untreated[ ,grep ("erlotinib", colnames(NCI_TPW_gep_treated))]

# normalise matrix with scale 
e_treated <- scale(e_treated, scale = TRUE)
e_untreated <- scale(e_untreated, scale = TRUE)

# change colnames according to their cell line instead of their complete sample name
cellline <- sapply(colnames(e_treated), function(x){
  annotation[x,"Cellline"]
  }) 
colnames(e_treated) <- cellline
colnames(e_untreated) <- cellline

# calculate foldchange by substracting e_untreated from e_treated
e_foldchange <- e_treated - e_untreated
```

### Table of cell lines with the highest variance   

To begin our cell line analysis, we want to identify those cell lines, which have the the highest variance of fold change. This means, that the fold changes of the various genes in those cell lines are spread over a wide range. Therefore, we calculate the variance of the fold change for each cell line and select the 15 cell lines with the highest variance ("table_celllines_var_top15"). 

```{r comment = NA, warning = FALSE}
# calculate the variance of each cell line
var_cellline <- apply(e_foldchange, 2, var)

# select 15 cell lines with highest variance (greater than 75% quantile, sorted by decreasing value)
cellline_var_greater_75quantile <- sort(var_cellline [which (abs(var_cellline) > quantile(abs(var_cellline), 0.75))], decreasing = TRUE)

# round variance to 5 digits
cellline_var_greater_75quantile <- round(cellline_var_greater_75quantile, digits = 5)

# create data frame with cell line names for top15 cell lines
celllines_top15 <- as.data.frame(names(cellline_var_greater_75quantile))

# add column with cancertype for top15 celllines
annotation_cancertype <- annotation[which(annotation[, "Drug"] == "erlotinib"), "Cancertype"]
names(annotation_cancertype) <- colnames(e_foldchange)
cancertypes_top15 <- sapply(names(cellline_var_greater_75quantile), function(x){
  annotation_cancertype[which(x == names(annotation_cancertype))]
})
table_celllines_var_top15 <- cbind(1:nrow(celllines_top15), celllines_top15, cellline_var_greater_75quantile, cancertypes_top15)

# change column and rownames of the table
colnames(table_celllines_var_top15) <- c("Rank", "Cellline", "Variance", "Cancertype")
rownames(table_celllines_var_top15) <- c(1:nrow(celllines_top15))

# load library formattable to create a table with our top 15 cell lines
library(formattable)
formattable(table_celllines_var_top15)
```

### Most divergent cell lines

In addition to the variance, we now perform a **PCA** to find those cell lines, which differ most from the other cell lines. In the following plot, those 15 cell lines are labeled and colored according to their cancertype to see whether cells from the same cancertype are clustered. 

```{r message = FALSE, warning = FALSE}
# draw PCA, where each point represents a cell line and is colored according to cancer type
pca <- prcomp(e_foldchange)
par(mar = c(4, 4, 4, 7))
plot(pca$rotation[,1], 
     pca$rotation[,2], 
     col = color_vector_cancertype, 
     pch = 19, 
     main = "PCA of cell lines",
     cex.main = 1.2,
     # get_eigenvalue is used to add percentage of explained variance in xlab and ylab
     # 2nd column of get_eigenvalue contains percentage of explained variance
     # round(x, digits to round): here round explained variance to one digit
     xlab = paste("PC1 (", round(get_eigenvalue(pca)[1, 2], 1), "%)"), 
     ylab = paste("PC2 (", round(get_eigenvalue(pca)[2, 2], 1), "%)"))

# add legend to associate each color with a cancertype
legend("topright", 
       legend = names(color_palette_cancertype), 
       col = color_palette_cancertype, 
       pch = 19, 
       xpd = TRUE, 
       inset = c(-0.18, 0))

# label cell lines, whose PC1 rotation value is greater than 0.14, representing the most divergent cell lines
celllines_PCA <- names(pca$rotation[ , 1])[pca$rotation[ , 1] > 0.14]
text(pca$rotation[ , 1][celllines_PCA], 
     pca$rotation[ , 2][celllines_PCA], 
     celllines_PCA,
     pos = 2,
     cex = 0.6)
```
Most of the cell lines identified by PCA as differing a lot from the rest are also part of the 15 cell lines showing the greatest variance. Furthermore, it can be seen that there is no clustering according to cancertype and that in the labeled cell lines nearly every cancertype is represented.

### Cell line clustering according to affected pathways

Lastly, we want to identify cell line clusters according to the regulation of different cellular pathways using a heatmap. For this aim we use two methods to create a heatmap. The **first heatmap** uses the **Supplementary Table S2** from the NCI TPW. In this table 4 to 16 genes are listed for each of the pathways. We first have to assign these genes to the pathways and can then compare the differences of the mean fold change of the genes of one pathway over all samples. Using an elbow plot, we try to find the optimal number of clusters. Thereafter, we can visualize the differences in a heatmap.  

```{r}
# assign genes to pathways (as in supplementary material)
MAPK <- c("AURKA", "AURKB", "SON", "CENPA", "KIF11", "DUSP6", "TPX2", "EGR1")
AKT_PI3K <- c("CFLAR", "XIAP", "BIRC5", "BIRC3", "GADD45A", "MCL1", "BCL2", 
              "BCL2L1", "HIF1A", "AR", "STAT3", "IL6", "JUN", "FOS")
  # AP1 for AKT_PI3K excluded (genename does not exist in FC)
  # same problem for VEGF
Cell_Cycle_Checkpoint <- c("CHEK1", "CHEK2", "TP53", "MDM2", "CDKN1A", "CCNE1", "CDK2", 
                           "CDC25A", "SMC1A", "CCNB1", "CDK1", "CDC25B", "CDC25C", 
                           "PLK1", "WEE1", "CCND1")
JAK_STAT <- c("SOCS1", "NMI", "BCL2L1", "CDKN1A", "MYC")
Immunity_Related <- c("BCL2L1", "XIAP", "BIRC5", "CFLAR", "IL6", "IL1B", "TNF")
  # PTGS does not exist
DNA_repair <- c("XRCC6", "PRKDC", "DCLRE1C", "XRCC4", "LIG4", "BRCA1", 
                "RAD52", "BRCA2", "RAD54L", "ATM", "ATR", "PARP1")
  # NHEJ1 does not exist
DNA_damage <- c("TP53", "FAS", "BAX", "PMAIP1")
  # BBC3 & ARG do not exist
ER_Stress_Survival <- c("HSPA5", "HSP90B1", "XBP1", "P4HB", "ATF4", "GADD45A")
ER_Stress_Apoptotic_Response <- c("ATF3", "DDIT3", "TNFRSF10B", "TRIB3", "BCL2L11", "CASP2")
  # BBC3 & GADD34 do not exist
Apoptosis_extrinsic_activation <- c("BID", "FAS", "CASP8", "CASP9", "APAF1")
Apoptosis_intrinsic_activation <- c("BAX", "BAK1", "BID", "PMAIP1", 
                                    "CASP6", "CASP9", "APAF1")
  # BBC3 does not exist
Autophagy_recycling_starvation <- c("MAP1LC3B", "ULK1", "ATG13", "BECN1", 
                                    "SQSTM1", "ATG5", "ATG12", "CTSB")
Autophagy_toxic <- c("SQSTM1", "ATG7", "RIPK1", "CTSB")

# create named list of all pathways
pathways <- list(MAPK = MAPK, 
                 AKT_PI3K = AKT_PI3K, 
                 Cell_Cycle_Checkpoint = Cell_Cycle_Checkpoint, 
                 JAK_STAT = JAK_STAT,
                 Immunity_Related = Immunity_Related,
                 DNA_repair = DNA_repair,
                 DNA_damage = DNA_damage,
                 ER_Stress_Survival = ER_Stress_Survival,
                 ER_Stress_Apoptotic_Response = ER_Stress_Apoptotic_Response,
                 Apoptosis_extrinsic_activation = Apoptosis_extrinsic_activation, 
                 Apoptosis_intrinsic_activation = Apoptosis_intrinsic_activation, 
                 Autophagy_recycling_starvation = Autophagy_recycling_starvation,
                 Autophagy_toxic = Autophagy_toxic) 
rm(MAPK, AKT_PI3K, Cell_Cycle_Checkpoint, JAK_STAT, Immunity_Related, DNA_repair, 
   DNA_damage, ER_Stress_Survival, ER_Stress_Apoptotic_Response, 
   Apoptosis_extrinsic_activation, Apoptosis_intrinsic_activation, 
   Autophagy_recycling_starvation, Autophagy_toxic)
```

After having assigned the genes to pathways, we create a "heatmap_data" matrix. We therefore calculate the mean fold change of all genes of one pathway for each cell line and each pathway.

```{r}
# calculate mean FC over genes of one pathway for each sample
# heatmap_data: 59 rows (samples), 13 columns (pathways)
heatmap_data <- sapply(1:length(pathways), function(x){ #for each pathway
  sapply(1:ncol(e_foldchange), function(y){ #for each cell line
    #calculate mean FC of genes of one pathway
    mean(e_foldchange[pathways[[x]], y]) 
  })
})

# change colnames to the names of the pathways and rownames to the names of the cell lines 
colnames(heatmap_data) <- c(names(pathways))
rownames(heatmap_data) <- colnames(e_foldchange)
```

To choose the appropiate amount of clusters, we perform an elbow plot. We thereby use the Within cluster sum of squares methode (WSS). This methode tries to minimize the within group distance and maximize the between-group distance to get compact clusters.

```{r,  fig.height = 3, warning = FALSE, message = FALSE}
# draw elbow plot to get optimal number of cell line clusters
library(factoextra)
fviz_nbclust(heatmap_data, 
             FUN = hcut, 
             method = "wss") + #within cluster sums of squares
  labs(title = "Elbow plot: optimal number of cell line clusters")

# draw elbow plot to get optimal number of pathway clusters
fviz_nbclust(t(heatmap_data), 
             FUN = hcut, 
             method = "wss") +
  labs(title = "Elbow plot: optimal number of pathway clusters")
```
The first elbow plot suggests to use two or three clusters for the cell lines. As the detection of the kink in the kurve is very subjective, we tried both possibilities, but the heatmap with three clusters looked better. For the second elbow plot, we choose three pathway clusters. 
  
The heatmap is created With a little white space between the three identified clusters for cell lines and pathways and an annotation bar showing the cancertype each sample belongs to is added at the top. To reduce the effect of outliers on the color scheme, we choose to color the heatmap according to quantiles.

```{r, fig.height = 6, warning = FALSE}
# load pheatmap and the color palette RcolorBrewer
library(pheatmap)
library(RColorBrewer)

# change annotation matrix into a data fram
annotation_cancertype_df <- as.data.frame(annotation_cancertype)
colnames(annotation_cancertype_df) <- "Cancertypes"

# scale pathways
heatmap_data <- scale(heatmap_data, scale = TRUE)

# color according to quantiles
# save quantiles from 0 - 100% in steps of 20%, used for defining the color breaks for 5 different colors
color_quantiles <- quantile(heatmap_data, seq(0, 1, 0.2))

# generate heatmap with colored bars according to clusters
pheatmap(t(heatmap_data),
         clustering_method = "ward.D2",
         annotation_col = annotation_cancertype_df,
         cutree_rows = 3, # makes white spaces between clusters
         cutree_cols = 3,
         main = "Affected pathways by erlotinib",
         mainsize = 1.2,
         show_colnames = FALSE,
         cellwidth = 4,
         cellheight = 15, 
         fontsize = 10,
         fontsize_row = 12,
         breaks = color_quantiles,
         col = rev(brewer.pal(5, "RdBu"))) # 5 colors from "RdBu" color palette
```


For the **second heatmap** we use the R package **PROGENy**, which scores 100 genes per pathway to calculate the pathway activity. This means that much more genes were used this time and our results should be more precise. 

```{r warning = FALSE}
# load progeny
library(progeny)

# create progeny heatmap with the fold change of all genes in all cell lines
progeny_heatmap <- progeny(e_foldchange)
```

Before creating the heatmap we again draw an elbow plot to determine how many clusters are needed. 

```{r, fig.height = 3, warning = FALSE}
# draw elbow plot to get optimal number of cell line clusters
library(factoextra)
fviz_nbclust(progeny_heatmap, 
             FUN = hcut, 
             method = "wss") + #within cluster sums of squares
  labs(title = "Elbow plot: optimal number of cell line clusters")

# draw elbow plot to get optimal number of pathway clusters
fviz_nbclust(t(progeny_heatmap), 
             FUN = hcut, 
             method = "wss") +
  labs(title = "Elbow plot: optimal number of pathway clusters")
```

In both elbow plots a kink at two clusters can be seen. Therefore we use two clusters for the cell lines and two clusters for the pathways.


Moreover, we create a dendrogramm to visualize the cluster structure in the heatmap and color the cell lines according to their cancer type. The fold change is once again colored according to quantiles. We thereby hope to see some patter to compare the cell lines. 

```{r warning = FALSE, fig.height = 6, message = FALSE}
# load dendextend
library(dendextend)

# create matrix called "cellline_clusters", with one column for cell line, cancer type and cluster number, to which the cell line belongs to in the heatmap
cellline_clustering <- hclust(dist(progeny_heatmap), method = "ward.D2")
cellline_groups <- as.data.frame(cutree(tree = as.dendrogram(cellline_clustering), k = 2))
cellline_clusters <- as.data.frame(cbind(cellline_groups, annotation_cancertype_df))
colnames(cellline_clusters) <- c("Cellline cluster", "Cancertype cluster")

# color according to quantiles
color_quantiles <- quantile(progeny_heatmap, seq(0, 1, 0.2))

# generate heatmap with colored bars according to clusters
par(mar = c(8, 2, 2, 2))
pheatmap(t(progeny_heatmap),
         clustering_method = "ward.D2",
         annotation_col = annotation_cancertype_df,
         cutree_rows = 2, #make white space between the different clusters
         cutree_cols = 2,
         main = "Affected pathways by erlotinib (PROGENy)",
         mainsize = 1.2, 
         show_colnames = FALSE,
         cellwidth = 5,
         cellheight = 20, 
         fontsize = 10,
         fontsize_row = 14,
         breaks = color_quantiles,
         col = rev(brewer.pal(5, "RdBu")))
```
Looking at both heatmaps, it is difficult to compare them. Unfortunately, apart from two or three exceptions, the pathways to which the genes are assigned, do not match. Moreover, the pathways for the first heatmap only include a small number of genes for each pathway and we simply took the mean fold change. The PROGENy heatmap on the other side takes around 100 genes per pathway and scores them, which is much more informative. Therefore, the clusters are also better visible. Furthermore,in the first heatmap, we have three clusters for the cell lines, while in the second one, we only have two. This means that the cell lines are in a different order complicating a comparison even more. 
We therefore decided to concentrate only on the PROGENy heatmap for further analysis, as this heatmap is much more precise and the clusters are better visible. Moreover, we focus only on the second cell line cluster (in the heatmap on the right side), because there EGFR and other proliferation promoting pathways are strongly downregulated.

**Downregulation of VEGF, PI3K, EGFR and MAPK**
  
  To begin, the erlotinib target EGFR is downregulated by the drug.

EGFR, also called epidermal growth factor receptor, is stimulated, as its name implies by epidermal growth factors. After its activation many different pathways are induced like cell proliferation, survival or growth. Erlotinib also binds to the EGFR and therefore blocks those signalling pathway inhibiting tumor progression.   

Another downregulated pathways is the MAP Kinase pathway, which is part of the Ras/Raf signalling cascade. The MAPK is dependent on EGFR signalling and phosphorylates transcription factors like c-myc, which promote again survival and proliferation. 

Moreover, the Phosphatidylinositol-3-Kinase (PI3K), which is part of the AKT signaling cascade, is also EGFR dependent. Akt signalling is associated with cell growth, apoptosis resistance, cell invation and also cell migration.  
Lastly, the vascular endothelial growth factors (VEGFs) are downregulated. They bind to the VEGF receptor and promote angiogenesis, a key pathway for tumor cells to get access to nutrients and oxygen from the blood. To conclude, all these different pathways, promoting tumor growth and survival are inhibited by the blocking of the epidermal growth factor receptor (EGFR) with erlotinib.(https://www.sigmaaldrich.com/technical-documents/articles/biology/egf-egfr-signaling.html) 

**Upregulation of tumor promoting pathways like Hypoxia and JAK.STAT**
  
   However, Erlotinib treatment also leads to an upregulation of the JAK.STAT and the Hypoxia pathway, which both promote tumor progression, oncogenesis and angiogenese in nearly every cell line. Cell lines in the left cluster, on the other hand, show nearly inverted results. (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5045092/)


### Venn Diagramm 

To sum up this milestone, we perfom a venn diagram to identify the cell lines, which are considered to be highly regulated by erlotinib in all three methods. We therefore, compare the results from the table with the highest variance, the PCA analysis and the PROGENy heatmap.

```{r message = FALSE, warning = FALSE}
# load package VennDiagram
library(VennDiagram)

# create vector, with cell lines of the second cluster from the PROGENy heatmap
heatmap_celllines_cluster2 <- rownames(cellline_clusters)[which(cellline_clusters[, 1] == 2)] 

# create list with cell lines of all three methods
venn.list <- list("Heatmap Cluster 2" = heatmap_celllines_cluster2,
                  "Top15 variance" = table_celllines_var_top15[, 2], 
                  "PCA" = celllines_PCA)

# draw venn diagram with cell lines of the highest variance table, PCA and heatmap cluster
venn.plot <- venn.diagram(
  x = venn.list,
  filename = NULL, 
  fill = c("blue", "red", "green"), 
  main = "Venn Diagramm of most regulated cell lines", 
  main.fontface = 2,
  main.cex = 1.2,
  main.fontfamily = "sans", # font type of title is Arial 
  cat.fontfamily = "sans",
  cat.cex = 0.9,
  fontfamily = "sans",
  cat.just = list(c(0.5, -40), c(0.2, -1), c(0,0)));
grid.draw(venn.plot);

# create vector, with ten cell lines of intersection
celllines_overlap <- calculate.overlap(venn.list)[[1]] 
```

With this plot, we reduce our cell lines to a total of 10, which were selected through all three different methods. The cell lines from the intersection are ACHN (Renal), CAKI-1 (Prostate), HCT-15 (Colon), IGR-OV1 (Leukemia), MDA-MB-468 (CNS), NCI-H322M (Breast), NCI-H522 (Breast), OVCAR-3 (Ovarian), SK-OV-3 (NSCLC) and SN12c (Renal). These cell lines will be used for further analysis and in combination with the biomarkers in milestone 5.

##   _**3. Milestone: Identify biomarkers**_

<div style = "text-align:justify">
As third milestone, we want to select those genes, which are most regulated by erlotinib. Therefore, we perform an MA plot to illustrate and identify genes with the highest fold change. Thereafter, we further reduce those genes according to their significance using a volcano plot. To illustrate, how many genes are left out due to their lower significance, we draw a venn diagram. Those genes, which have a high fold change and a high significance represent our biomarkers. Selecting the biomarkers, we draw a density plot to compare the gene expression before and after treatment. 

### Select biomarkers with high foldchange
Firstly, we draw an **MA plot** to compare the fold change to the average log2 expression of all genes. Thereby, we select only those genes, which have a mean fold change higher than 0.25 or a mean fold change lower than -0.25.
```{r message = FALSE, warning=FALSE}
# load ggplot2 and ggrepel
library(ggplot2)
library(ggrepel)

# create data frame with variables M and A of MA-plot as well as one column, which shows, if absolute fold change is greater than 0.25 
# M = log2(treated) - log2 (untreated)
M <- e_foldchange 
# average log2-expression value A = 1/2 (log2(treated)+log2(untreated))
A <- 0.5 * (e_treated + e_untreated) 
MA <- cbind("M" = rowMeans(M), "A" = rowMeans(A))
rm(M, A)
MA <- as.data.frame(MA)
MA$High_foldchange <- ifelse(MA$M > 0.25 | MA$M < -0.25, "FC_abs > 0.25", "FC_abs < 0.25")

# label genes of MA plot, which have an absolute foldchange greater than 0.5 
MA_labeled <- MA[which(abs(MA[ , "M"]) > 0.5) , ]

# create MA plot, where genes with FC_abs > 0.25 are colored and those genes with FC_abs > 0.5 are labeled
ggplot(data = MA)+ 
  aes(x = A, y = M, color = High_foldchange)+
  geom_point()+
  xlab("average log2 expression")+
  ylab("log fold change")+
  ggtitle("MA plot of all genes")+
  geom_text(data = MA_labeled, aes(A, M, label = rownames(MA_labeled)), hjust = 0, nudge_x = 0.1)+
  # change the size and face (bold) of the title and center it (hjust = 0.5)
  theme(plot.title = element_text(size = 13, face = "bold", hjust = 0.5)) 

#save highlighted biomarkers, obtained through MA plot in biomarkers_MA
biomarkers_MA <- rownames(MA)[which(abs(MA$M) > 0.25)]
```

### Select biomarkers with high fold change and high significance
We then create a **volcano plot** to find the genes, which have in addition to a high fold change, also a very high significance. With a paired two-sample t-test between the untreated and treated gene expression values a p-value was calculated for each gene. These p-values were then corrected with the Benjamini-Hochberg procedure to get the so-called FDR-values, which are used as a measurement for the significance of a fold change.
```{r warning = FALSE, message = FALSE}
# load libraries needed for EnhancedVolcano
library(ggplot2)
library(ggrepel)
library(EnhancedVolcano)

# calculate mean foldchange over all cell lines for each gene
e_foldchange_mean_over_cell_lines <- rowMeans(e_foldchange) 

# determine p-value for a paired two-sample t-test between gene expression of treated and untreated samples
p_values <- sapply(rownames(e_treated), function(x) {
  # perform t-test and save p-values of each gene in p_values-vector
  t.test(e_treated[x,], e_untreated[x,],paired = T)$p.value}) 

# calculate FDR with benjamini-hochberg (BH)
# selecting significance of genes with p-value, would give us more false-positives, because most of the genes are not regulated by erlotinib
FDR_values <- p.adjust(p_values, method = "BH", n = length(p_values)) 

# create matrix with one column for mean foldchange and one column for FDR_values for every gene
volcano_plot_mat <- cbind(e_foldchange_mean_over_cell_lines, FDR_values)

# draw volcano plot, where those genes with FC_abs > 0.25 and FDR_value > 10e^-9 are labeled and highlighted in red 
EnhancedVolcano(volcano_plot_mat, 
                lab = rownames(volcano_plot_mat),
                x = "e_foldchange_mean_over_cell_lines", 
                y = "FDR_values",
                title = "Volcano plot of all genes",
                titleLabSize = 13,
                ylab = bquote(~FDR~value),
                pCutoff = 10e-9, # threshold for coloring significant ones
                FCcutoff = 0.25, # threshold for coloring high FC
                transcriptPointSize = 2,
                transcriptLabSize = 3,
                legendPosition = "right",
                legendLabSize = 10,
                axisLabSize = 10)

# save highlighted biomarkers, obtained through volcano plot in biomarkers_volcano
biomarkers_volcano <- rownames(volcano_plot_mat)[which(abs(volcano_plot_mat[ , 1]) >  0.25  
                                                       & volcano_plot_mat[ , 2] < 10e-9)]
```


### Compare selected biomarkers 
We use a **Venn Diagram** to compare the most regulated genes obtained through the MA plot with those obtained through the volcano plot. This shows us, how many genes were sorted out due to a lower FDR_value.
```{r warning = FALSE, message = FALSE}
# draw Venn Diagram with biomarkers of MA and Volcano plot
par(oma = c(2, 2, 2, 2))
venn.plot <- venn.diagram(
  x = list(
    "Volcano Plot" = biomarkers_volcano,
    "MA Plot" = biomarkers_MA),
  filename = NULL, 
  fill = c("blue", "red"), 
  main = "Venn Diagramm of selected biomarkers", 
  main.fontface = 2,
  main.cex = 1.2,
  main.fontfamily = "sans", # font type of title is Arial 
  cat.fontfamily = "sans",
  cat.cex = 0.9,
  fontfamily = "sans",
  ext.text = FALSE,
  cat.just = list(c(-0.8,2), c(1,0)));
grid.draw(venn.plot);
```

### Compare distribution of gene expression of biomarkers before and after treatment  
To sum up the milestone, we draw a **density plot** only with the biomarkers identified in the volcano plot. One line is drawn to show the distribution of all untreated values (in green) and another one to show the distribution of the treated ones (in red) to show how the overall gene expression of the biomarkers changed due to erlotinib treatment. To visualize the differences in the distribution, we have to separate between up- and downregulated biomarkers to make sure that positive and negative fold changes do not cancel each other out.
```{r}
# create new vector, where only up- or downregulated volcano biomarkers are included
biomarkers_volcano_up <- rownames(volcano_plot_mat)[which(volcano_plot_mat[ , 1] > 0.25)]
biomarkers_volcano_down <- rownames(volcano_plot_mat)[which(volcano_plot_mat[ , 1] < -0.25)]

# density plot with upregulated biomarkers_volcano to compare untreated with treated gene expression
plot(density(e_untreated[biomarkers_volcano_up, ]), 
     "Density plot of gene expression of upregulated biomarkers", 
     cex = 1.2, 
     col = "green", 
     ylim = c(0.0, 0.5))
lines(density(e_treated[biomarkers_volcano_up, ]), col = "red")
legend("topright", legend = c("untreated", "treated"), col = c("green", "red"), pch = 15)

# density plot with downregulated biomarkers_volcano to compare untreated with treated gene expression
plot(density(e_untreated[biomarkers_volcano_down, ]), 
     "Density plot of gene expression of downregulated biomarkers", 
     cex = 1.2, 
     col = "green", 
     ylim = c(0.0, 0.7))
lines(density(e_treated[biomarkers_volcano_down, ]), col = "red")
legend("topright", legend = c("untreated", "treated"), col = c("green", "red"), pch = 15)
```
To conclude this milestone, we end up with 63 biomarkers, which have an absolute fold change greater than 0.25 and a FDR-value greater than 10e^-9. The density plots created of the biomarkers show a considerable difference between the untreated and treated gene expression, since all the rarely regulated genes are left out. 


##   _**4. Milestone - Does the gene expression correlate with the drug sensitivity?**_
<div style = "text-align:justify">
In the fourth milestone, we focus on the erlotinib target EGFR and its regulated gene EGR1. The GI50-values of all cell lines, which show, how much the cell growth is inhibited and therefor how sensitive the cell line is towards erlotinib, were compared with the expression of these genes. Thereby, we hope to see a correlation between the untreated gene expression of EGFR and the GI50 values, because the erlotinib receptor EGFR should be more expressed in erlotinib sensitive cells. Thereafter, we correlate the fold change of EGR1 with the GI50-values, since we expect, that the fold change of EGR1 decreases most in erlotinib sensitive cell lines.

### Data preparation 
```{r}
# create vector of NegLogGI50 of erlotinib, which only includes cell lines, which also appear in e_foldchange
NegLogGI50_59_celllines <- NegLogGI50 ["erlotinib", -c(8,29)] # remove cell line "COLO-205" and "MDA-N"

# multiply the NegLogGI5 = by -1 to get the LogGI50-values
LogGI50_59_celllines <- NegLogGI50_59_celllines * (-1)


# change colnames of e_untreated with cell line instead of complete sample name
colnames(e_untreated) <- cellline

# create color vector to color erlotinib treated cell lines according to cancertype
e_color_cancertype <- color_vector_cancertype[grep("erlotinib", names(color_vector_cancertype), value = TRUE)]
```

### EGFR gene expression in the untreated cell lines against GI50-values   
Firstly, we draw a **scatter plot** to show, if the expression level of EGFR without treatment correlates with the sensitivities of the cell lines. Performing a Pearson correlation, we hope to get a correlation value close to -1, since we would expect that the higher the EGFR expression, the lower is the GI50 value. Moreover, we plot a linear regression line, to see how well the line fits to the individual points and if we can deduce the sensitivity of the cell line from the untreated gene expression of EGFR. 
```{r pressure, warning = FALSE}
# draw scatter plot of  untreated gene expression level of EGFR aginst LogGI50 and color them according to the cancer type
par(mar = c(4, 4, 4, 7), xpd = "TRUE")
plot(LogGI50_59_celllines, e_untreated ["EGFR",], 
      col = e_color_cancertype, 
      pch = 19, 
     xlab = "logGI50", 
     ylab = "EGFR Expression (untreated)",        
     main = "Expression of the epidermal growth factor receptor (Her 1)",
     cex.main = 1.2)

# add linear regression line
# set clipping limit that abline is not longer than the plot itself 
clip(-7.13, -3.88, -1, 2) 
abline(lm (e_untreated["EGFR", ]~ LogGI50_59_celllines))

# add legend to associate each color with a cancertype
legend("topright", 
    legend = names(color_palette_cancertype), 
    col = color_palette_cancertype, 
    pch = 19,
    inset = c(-0.18, 0),
    xpd = TRUE)

# label cell lines, which are most sensitive and have a high e_untreated expression of EGFR 
labeled_celllines_EGFR <- names(LogGI50_59_celllines)[LogGI50_59_celllines < - 5.5
                                                        & e_untreated ["EGFR", ] > 0]
text(LogGI50_59_celllines[labeled_celllines_EGFR], e_untreated ["EGFR", labeled_celllines_EGFR], 
     labels = labeled_celllines_EGFR,
     cex = 0.7,
     pos = 3)

# perform Pearson correlation, to see how well LogGI50 and the foldchange correlate with each other 
cor_EGFR <- cor.test(LogGI50_59_celllines, e_foldchange["EGFR",], method = "pearson")
cor_EGFR_estimate <- round(cor_EGFR$estimate, digits = 3)

# add title to scatter plot, which includes the correlation value
cor_text <- paste("r = ", cor_EGFR_estimate)
mtext(side = 1, line = -1.2, text = cor_text, adj = c(0.01))
```
Unfortunately, the correlation value is only -0.352 and especially the cell lines with a lower untreated gene expression do not correlate well with the sensitivity of the cell line. However, most of the labeled cell lines accumulate close to the regression line. 

### EGR1 foldchange against GI50-values 
A **scatter plot** to plot the fold change of EGR1 against the GI50-values was drawn, as this gene is regulated by EGFR. EGR1 is a transcriptional factor and is assocciated with the activiation of tumor suppressor genes like p53/TP53 and TGFB1, and plays an important role in the regulation of growth factor responses. This plot with EGR1 was also drawn for 6h in the supplementary data provided by NCI Transcriptional Pharmacodynamics Workbench. We therefore, want to compare this with our scatter plot after 24h.

```{r figEGR1 , fig.show = "hold",  fig.width = 7, fig.height = 7, warning = FALSE,  message = FALSE}
# import picture shown in the Supplementary after 6h
url <- "https://raw.githubusercontent.com/datascience-mobi/project-02-group-04/master/Project%20proposal/EGR1%20expression.PNG"
include_graphics(url)

# create new color vector, which has the same colors as the imported picture with GI50 scatter plot of 6h
color_palette_cancertype_EGR1 <- c("red", "chartreuse", "orange", "green3", "turquoise", "hotpink", "blue", "purple", "deepskyblue")
names(color_palette_cancertype_EGR1) <- levels(cellline_annotation$Cancer_type)
color_vector_cancertype_EGR1 <- sapply(rownames(annotation), function(x){
  unname(color_palette_cancertype_EGR1[annotation[x, 3]]) })
e_color_cancertype_EGR1 <- color_vector_cancertype_EGR1[grep("erlotinib", names(color_vector_cancertype_EGR1), value = TRUE)]

# draw scatter plot of foldchange of EGR1 against LogGI50 and color them according to the cancer type
par(mar = c(4, 4, 4, 10), xpd = "TRUE")
plot(LogGI50_59_celllines, e_foldchange ["EGR1",], 
     col = e_color_cancertype_EGR1, 
     pch = 19, 
     xlab = "logGI50", 
     ylab = "EGR1 Expression (log2, relative to control)",        
     main = "Erlotinib 24h", 
     cex.main = 1.2, 
     col.main = "deepskyblue")

# add legend to associate each color with a cancertype
segments(-7.11, 0, -3.88)
legend("topright", 
       legend = names(color_palette_cancertype_EGR1), 
       col = color_palette_cancertype_EGR1, 
       inset = c(-0.3, 0),
       pch = 19)


# label cell lines, which are most sensitive and have a very low foldchange of EGR1 
labeled_celllines <- names(LogGI50_59_celllines)[LogGI50_59_celllines < - 6
                                                        & e_foldchange["EGR1", ] < 0 
                                                 | LogGI50_59_celllines < -5 
                                                        & e_foldchange["EGR1", ] < -3]
text(LogGI50_59_celllines[labeled_celllines], e_foldchange ["EGR1", labeled_celllines],
     labels = labeled_celllines,
     cex = 0.7,
     pos = 4)

# add linear regression line
# set clipping limit that abline is not longer than the plot itself
clip(-7.13, -3.88, -2, 0.5)  
abline(lm (e_foldchange["EGR1", ]~ LogGI50_59_celllines))

# perform Pearson correlation, to see how well LogGI50 and the foldchange correlate with each other 
cor_EGR1 <- cor.test(LogGI50_59_celllines, e_foldchange["EGR1",], method = "pearson")
cor_EGR1_estimate <- round(cor_EGR1$estimate, digits = 3)

# add title to scatter plot, which includes the correlation value
cor_text <- paste("r = ", cor_EGR1_estimate, "(***)")
mtext(side = 3, line = -1.2, text = cor_text, adj = c(0.01))

```
This time, the correlation value is much higher with r = 0.606 and apart from a few outliers, the sensitivity of the cell lines correlates quite well with the EGR1 fold change. Comparing 6 and 24h, the same correlation trend is seen in both scatter plots. However, the fold change of some of the most sensitive cell lines as "MDA-MB-468" or "IGR-OV1" changed dramatically between the two time points. 

##   _**5. Milestone - final results**_

<div style = "text-align:justify">
In our last milestone, we further reduce our cell lines of milestone 2 by selecting only the erlotinib sensitive cell lines. Moreover, we further reduce the biomarkers by selecting only the biomarkers, which are most regulated in erlotinib sensitive cell lines. With these final biomarkers and cell lines we draw a boxplot, a foldchange-GI50 correlation plot and one single gene response plot for each biomarker to compare the changes in gene expression after erlotinib treatment and sum up our project. 

### Select final cell lines according to GI50 correlation
In Milestone 2, 10 cell lines were selected from the intersection of the cell lines with the highest variance, the cell lines gained by PCA and the cell lines, which formed cluster 2 in the heatmap. We now want to compare those cell lines with the 10 cell lines, which have the highest LogGI50 and are therefore most sensitive to erlotinb using a  **Venn Diagram**. 

```{r message = FALSE}
library(VennDiagram)

# save the names of the 10 most erlotinib-sensitive cell lines (highest LogGI50)
LogGI50_top10 <- names(sort(LogGI50_59_celllines)[1:10])

# Venn diagram to compare cell lines from Milestone 2 with the ones with highest sensitivity
venn.plot <- venn.diagram(
  x = list(
    "Overlap of variance, 
    PCA and heatmap cell lines" = celllines_overlap,
    "Top 10 LogGI50" = LogGI50_top10),
  filename = NULL, 
  fill = c("blue", "red"), 
  main = "Venn Diagramm of most regulated cell lines", 
  main.fontface = 2,
  main.cex = 1.2,
  main.fontfamily = "sans", # font type of title is Arial 
  cat.fontfamily = "sans", # font type of labels
  cat.cex = 0.9,
  fontfamily = "sans", 
  cat.just = list(c(0.2,-1.5), c(1.2,-4))); #change position of the labels
grid.draw(venn.plot);

#save the 7 cell lines from the intersection
celllines_final <- intersect(celllines_overlap, LogGI50_top10)

```
With this plot we can compare the 10 cell lines, which are highly upregulated, with the 10 cell lines, which are sensitive to erlotinib. Thereby, 7 final cell lines are identified, which are represented in both groups. These cell lines are ACHN (Renal), CAKI-1 (Prostate), IGR-OV1 (Leukemia), MDA-MB-468 (CNS), NCI-H322M (Breast), NCI-H522 (Breast) and SK-OV-3 (NSCLC). 

### Select final biomarkers according to GI50 correlation
Additionally, we want to further reduce the amount of biomarkers, we selected in Milestone 3 through the MA and volcano plot. This time we want to select only those genes, which are most regulated in erlotinib sensitive and less regulated in insensitive cell lines. We therefore select 500 genes, which have the highest correlation value of the fold change against the GI50 value. Drawing a **venn diagram**, we select those biomarkers, which belong to the volcano biomarkers representing a high foldchange and a high significance, but which also belong to the 500 genes with the highest absolute fold change in erlotinib sensitive cell lines. 
```{r}
# create vector, which contains the GI50 - fold change correlation value for all genes
cor_all_genes <- sapply(rownames(e_foldchange), function(x){
  cor_test_one_gene <- cor.test(LogGI50_59_celllines, e_foldchange[x,], method = "pearson")
  cor_test_one_gene$estimate
})
# sort genes according to decreasing correlation value and select 500 genes with the highest correlation value
cor_all_genes_sorted <- sort(abs(cor_all_genes), decreasing = TRUE)
cor_top_500_genes <- cor_all_genes_sorted[1:500]

# create vector, which contains the GI50 - fold change correlation value for the volcano biomarkers
cor_biomarkers <- sapply(biomarkers_volcano, function(x){
  cor_test_one_biomarker <- cor.test(LogGI50_59_celllines, e_foldchange[x,], method = "pearson")
  cor_test_one_biomarker$estimate
})

# plot venn diagram to find biomarkers, which belong to the volcano biomarkers and to the 500 genes with highest GI50 correlation value 
par(oma = c(2, 2, 2, 2))
biomarkers_MA_vector <- rownames(MA_labeled)
venn.plot <- venn.diagram(
  x = list(
    "500 genes with highest correlation" = names(cor_top_500_genes),
    "correlation of biomarkers" = names(cor_biomarkers)),
  filename = NULL, 
  fill = c("blue", "red"), 
  main = "Venn Diagramm of genes with highest GI50-foldchange correlation compared to biomarkers", 
  main.fontface = 2,
  main.cex = 1.2,
  main.fontfamily = "sans", # font type of title is Arial 
  cat.fontfamily = "sans",
  cat.cex = 0.9,
  ext.text = FALSE,
  fontfamily = "sans",
  cat.pos = c(0,0));
grid.draw(venn.plot);

# save final biomarkers, which are in intersect of volcano biomarkers and cor_top_500_genes
biomarkers_final <- intersect(names(cor_top_500_genes), names(cor_biomarkers))
biomarkers_final <- gsub(pattern = ".cor", replacement = "", x = biomarkers_final)
```
With this plot we can compare the 63 biomarkers, which have a high foldchange and a high significance, with the 500 genes with the highest correlation. Thereby, we find 5 final biomarkers, which are represented in both groups. These biomarkers are ERCC6L, ATAD2, E2F8, MKI67 and MIR15A//DLEU2L//DLEU2.

### Compare gene expression of final biomarkers 
To get a better overview of our five final biomarkers, a **boxplot** of the **fold change** was created. This shows us the median as well as the distribution of the fold changes between the different cell lines. 
```{r}
# create matrix only with fold change of biomarkers
e_foldchange_biomarkers_final <- sapply(biomarkers_final, function(x){
  e_foldchange[x, ]
})

# draw boxplot of fold change
par(mar = c(13, 4, 2, 2))
boxplot(e_foldchange_biomarkers_final, 
        ylab = "foldchange", 
        main = "Boxplot of foldchange of the volcano biomarkers", 
        cex = 1.2, 
        las = 2)
```
Thereby, we can see that the median fold change decreased in all our biomarkers and accounts for around -0.25.

To compare the gene expression before and after the treatment, we draw a **boxplot** of the **untreated vs. treated gene expression** of the biomarkers. 
```{r}
# create matrix only with gene expression of untreated and treated final biomarkers
e_untreated_biomarkers_final <- sapply(biomarkers_final, function(x){
  e_untreated[x, ]
}) 
e_treated_biomarkers_final <- sapply(biomarkers_final, function(x){
  e_treated[x, ]
})

# add "treated" to colnames of gene expression of treated biomarkers to distinguish them from untreated in a common matrix
colnames(e_treated_biomarkers_final) <- paste(colnames(e_treated_biomarkers_final),"Treated",sep = "_") 

# create matrix, which contains gene expression of untreated and treated 
e_treated_untreated_biomarkers_final <- cbind (e_treated_biomarkers_final, e_untreated_biomarkers_final)

# sort matrix according to colnames that treated and untreated gene expression of the same gene are next to each other
e_treated_untreated_biomarkers_final <- e_treated_untreated_biomarkers_final[,order(colnames(e_treated_untreated_biomarkers_final))]

# create color vector, where untreated samples are green and treated ones are red
color_boxplot_e_treated_untreated <- sapply(colnames(e_treated_untreated_biomarkers_final), function(x) {
  ifelse(x %in% grep ("Treated",colnames(e_treated_untreated_biomarkers_final), value = TRUE),"red", "green")})

# draw boxplot, where treated and untreated are right next to each other
par(mar = c(16, 4, 2, 2))
boxplot(e_treated_untreated_biomarkers_final, 
        ylab = "gene expression (log2)", 
        main = "Boxplot of gene expression of the volcano biomarkers", 
        ceex = 1.2, 
        las = 2, 
        col = color_boxplot_e_treated_untreated)
```
Looking at the boxplot, we can say, that the overall gene expression of ATAD2 and MKI67 are slightly higher than the gene expression of the other three genes in the untreated and treated samples. Moreover, in all genes the range of the gene expression levels is higher in the treated samples. This could be expected, as the genes are regulated differently in all of the 59 cell lines. 

### Foldchange- GI50 correlation of final biomarker
Moreover, we draw once again a **scatter plot** with the foldchange - GI50 correlation, but this time of our final biomarkers. We thereby want to compare if the fold changes for each cell line vary a lot between the five genes or if you can even see some pattern.   
```{r warning = FALSE}
# create correlation matrix with 5*59 rows and 4 columns (foldchange, GI50 value, color and biomarker name)
# vector of fold change, where the 59 fold change values for each of the five biomarkers are combined below each other
correlation_foldchange <- c(e_foldchange[biomarkers_final [1],],
                            e_foldchange[biomarkers_final [2],],
                            e_foldchange[biomarkers_final [3],],
                            e_foldchange[biomarkers_final [4],],
                            e_foldchange[biomarkers_final [5],])

# vector of GI50, where the 59 LogGI50 are repeated five times for the five biomarkers
correlation_GI50 <-rep(LogGI50_59_celllines, 5)

# vector of color, where each color is assigned to the 59 rows of one biomarker
correlation_color <- c(rep("green", 59), 
                       rep("yellow", 59),
                       rep("red", 59),
                       rep("blue", 59),
                       rep("black", 59))

# vector of biomarker name, where each biomarker name is assigned to the 59 rows of one biomarker
correlation_name_biomarker <- c(rep("ATAD2", 59),
                                rep("E2F8", 59),
                                rep("ERCC6L", 59),
                                rep("MIR15A//DLEU2L//DLEU2", 59),
                                rep("MKI67", 59))

# create matrix of the four vectors 
correlation_biomarkers <- cbind (correlation_foldchange, 
                                 correlation_GI50, 
                                 correlation_color, 
                                 correlation_name_biomarker)
rm(correlation_foldchange, correlation_GI50, correlation_name_biomarker)

# create color palette for each of the five biomarkers
color_palette_biomarkers <- c("green","yellow","red" , "blue", "black" )
names(color_palette_biomarkers) <- biomarkers_final

# draw scatter plot of foldchanges of the biomarkers aginst LogGI50 
par(mar = c(4, 6, 2, 13), xpd = "TRUE")
plot(correlation_biomarkers[ ,"correlation_GI50"],
     correlation_biomarkers[ ,"correlation_foldchange"],
     col = correlation_color, 
     pch = 19,
     xlab = "logGI50", 
     ylab = "foldchange of biomarkers",        
     main = " Correlation between biomarkers and the GI50 values of the different cell lines",
     cex.main = 1.2)

# add legend to associate each color with one final biomarker
legend("topright", 
       legend = names( color_palette_biomarkers),
       col = color_palette_biomarkers,
       pch = 19,
       inset = c(-0.5, 0),
       xpd = TRUE)
```
Looking at the scatter plot, we can see, that the fold changes of the five biomarkers do not vary a lot for less sensitive cell lines. However, for erlotinib sensitive cell lines the foldchanges differ dramatically at some points. Mostly the fold change of the MIR15A gene varies from the other four genes in the sensitive cell lines.

### Single gene response plot 
We therfore want to take a closer look at the fold change of the biomarkers in the most sensitive cell lines using a **barplot**. 
```{r message = FALSE, warning = FALSE}
#comparison with lapatinib
# new matrix only with samples treated with lapatinib  (l = lapatinib)
l_treated <- NCI_TPW_gep_treated[ ,grep ("lapatinib", colnames(NCI_TPW_gep_treated))]
l_untreated <- NCI_TPW_gep_untreated[ ,grep ("lapatinib", colnames(NCI_TPW_gep_treated))]

# normalise matrix with scale 
l_treated <- scale(l_treated, scale = TRUE)
l_untreated <- scale(l_untreated, scale = TRUE)

# change colnames according to their cell line instead of their complete sample name
cellline <- sapply(colnames(l_treated), function(x){
  annotation[x,"Cellline"]
  }) 
colnames(l_treated) <- cellline
colnames(l_untreated) <- cellline

# calculate foldchange by substracting e_untreated from e_treated
l_foldchange <- l_treated - l_untreated

lapatinib_ERCC6L <- l_foldchange[biomarkers_final[1], celllines_final]



# create data frame, with one column for the foldchange of each biomarker in the final cell lines
df_barplot_results <- as.data.frame(sapply (biomarkers_final, function(x) {
  e_foldchange[x, celllines_final]
}))

# add one column for the cell line and one for the cancer type of the cell line
df_barplot_results <- as.data.frame(cbind(cellline = celllines_final, annotation_cancertype_df[celllines_final, ], df_barplot_results, lapatinib_ERCC6L = lapatinib_ERCC6L))
colnames(df_barplot_results)[2] <- "Cancertypes" #change colname of cancer type column

# sort cell lines according to cancer type
df_barplot_results <- df_barplot_results [order(df_barplot_results[,"Cancertypes"]),]

# create barplot for each gene and color according to cancer type
plot <- lapply(colnames(df_barplot_results)[3:8], function(gene){
  ggplot_title <- paste("Foldchange of" , gene)
  ggplot(data = df_barplot_results, aes(x = cellline, y = df_barplot_results[ ,gene], fill = Cancertypes)) +
    geom_bar(stat = "identity")+
    scale_fill_manual(values = c("aquamarine", "forestgreen", "chartreuse", 
                               "cadetblue","purple", 
                               "firebrick1", "deepskyblue"))+
    ggtitle(paste0 (ggplot_title, "\n"))+
    theme(axis.title.x = element_blank())+
    theme(axis.title.y = element_blank())+
    scale_x_discrete(limits = rownames(df_barplot_results))+
    coord_flip()
})

# load ggpubr
library(ggpubr)
# draw all five barplots next to each other
ggarrange( plot[[2]], plot[[3]], plot[[4]], plot[[5]], plot[[1]], plot[[6]],
          ncol = 2, nrow = 3, common.legend = TRUE, legend = "top")
```

```{r}
# create data frame, with one column for the foldchange of each biomarker in the final cell lines
df_barplot_results <- as.data.frame(sapply (biomarkers_final, function(x) {
  l_foldchange[x, celllines_final]
}))

# add one column for the cell line and one for the cancer type of the cell line
df_barplot_results <- as.data.frame(cbind(cellline = celllines_final, annotation_cancertype_df[celllines_final, ], df_barplot_results))
colnames(df_barplot_results)[2] <- "Cancertypes" #change colname of cancer type column

# sort cell lines according to cancer type
df_barplot_results <- df_barplot_results [order(df_barplot_results[,"Cancertypes"]),]

# create barplot for each gene and color according to cancer type
plot <- lapply(colnames(df_barplot_results)[3:7], function(gene){
  ggplot_title <- paste("Foldchange of" , gene)
  ggplot(data = df_barplot_results, aes(x = cellline, y = df_barplot_results[ ,gene], fill = Cancertypes)) +
    geom_bar(stat = "identity")+
    scale_fill_manual(values = c("aquamarine", "forestgreen", "chartreuse", 
                               "cadetblue","purple", 
                               "firebrick1", "deepskyblue"))+
    ggtitle(paste0 (ggplot_title, "\n"))+
    theme(axis.title.x = element_blank())+
    theme(axis.title.y = element_blank())+
    scale_x_discrete(limits = rownames(df_barplot_results))+
    coord_flip()
})

# load ggpubr
library(ggpubr)
# draw all five barplots next to each other
ggarrange( plot[[1]], plot[[2]], plot[[3]], plot[[4]], plot[[5]], 
          ncol = 2, nrow = 3, common.legend = TRUE, legend = "top")
```

We sum up our specific analysis with this barplot. It shows the five final biomarkers, which have a high fold change, a high significance and a specific regulation only in erlotinib sensitive cell lines. The seven selected cell lines on the other hand show a high variance, differ most from the other cell lines in the PCA, build a cluster in the heatmap and are most sensitive to erlotinib. Comparing the five barplots, we can see that ERCC6L, ATAD2, E2F8 and MKI67 are mostly regulated in the ovarian cancer cell lines IGR-OV1 and SK-OV-3 and the renal cancer line ACHN. MIR15A, on the other hand, is not regulated a lot in the ovarian cancer lines, but in the renal cell lines ACHN, CAK-1 and the NSCLC cell line NCI-H322M. 
</div>